File: coding_challenges/candy_crush_1d/python/candy_crush_1d.py
"""Candy crush 1D module.
"""

from typing import List


class CandyNode:
    """Candy node class.
    """

    def __init__(self, candy: int) -> None:
        self.candy: str = candy
        self.amount: int = 1


class CandyCrush1D:
    """Find prefix class.
    """

    def __init__(self) -> None:
        pass

    def crush(self, candies: str) -> str:
        """Crush the candies received.

        Args:
            candies (str): The candies.

        Returns:
            str: Remaing candies.
        """
        candies_stack: List[CandyNode] = []

        # Add candies and crush them if possible.
        for candy in candies:
            # If empty or the top one is the same, add it and continue.
            if not len(candies_stack):
                candies_stack.append(CandyNode(candy))
                continue
            elif candies_stack[0].candy == candy:
                candies_stack[0].amount += 1
                continue

            # Crush the candies at the top if possible.
            while len(candies_stack) != 0 and candies_stack[0].amount >= 3:
                candies_stack = candies_stack[1:]

            # Add the candy at the top if exist, otherwise at the end.
            if len(candies_stack) and candies_stack[0].candy == candy:
                candies_stack[0].amount += 1
            else:
                candies_stack = [CandyNode(candy)] + candies_stack[:]

        # Crush candies left after adding them.
        while len(candies_stack) != 0 and candies_stack[0].amount >= 3:
            candies_stack.pop()

        # if no candy left, return empty result.
        if not len(candies_stack):
            return "-"

        # return remaing candies as a string.
        remain_candies = [candy.candy *
                          candy.amount for candy in candies_stack]
        remain_candies.reverse()
        return "".join(remain_candies)

File: coding_challenges/candy_crush_1d/cpp/src/CandyCrush1D.cpp
/*
 * Copyright 2023 Elías Serrano. All rights reserved.
 * License: https://github.com/FriendlyCodingPals/judgment-of-paris#license
 */

#include "CandyCrush1D.h"

#include <iostream>
#include <stack>

struct Candy {
  char type;
  int amount;
};

std::string CandyCrash1D::GreedyCrush(const std::string& candies) {
  std::stack<Candy> candies_stack;

  for (const auto& candy : candies) {
    if (candies_stack.empty()) {
      candies_stack.emplace(candy, 1);
      continue;
    } else if (candies_stack.top().type == candy) {
      ++candies_stack.top().amount;
      continue;
    }

    while (!candies_stack.empty() && candies_stack.top().amount >= 3) {
      candies_stack.pop();
    }

    if (!candies_stack.empty() && candies_stack.top().type == candy) {
      ++candies_stack.top().amount;
    } else {
      candies_stack.emplace(candy, 1);
    }
  }

  while (!candies_stack.empty() && candies_stack.top().amount >= 3) {
    candies_stack.pop();
  }

  std::string remain_candies = "";
  while (!candies_stack.empty()) {
    const auto& candy_stack_top = candies_stack.top();
    candies_stack.pop();

    std::string candy_stack_top_str = std::string(candy_stack_top.amount, candy_stack_top.type);
    remain_candies = candy_stack_top_str + remain_candies;
  }

  return remain_candies;
}
